### DISCUSSION TITLE
polling-xhr.js:157 POST https://cet-chat.herokuapp.com/socket.io/?EIO=4&transport=polling&t=NsGVGZJ&sid=2I5zXjHN7fUZN5lNAAAA 400 (BAD REQUEST)

### DISCUSSION THREAD

**Discussion by somesh1234567** [socketio/socket.io#4203] [Q&A]
``Hi everybody I am trying to build  a chat app using flask and socketio
The app is working fine in localserver(PORT: 5000)
but when I deploy it using gunicorn it is throing error
PROCFILE:  web: gunicorn -k geventwebsocket.gunicorn.workers.GeventWebSocketWorker -w 1 app:app
app.py:
`import os
from logging import debug
from flask import Flask, render_template, request, redirect, url_for, session
from flask_sqlalchemy import SQLAlchemy
from flask_socketio import SocketIO, join_room, leave_room, emit, rooms
from flask_session import Session

app = Flask(__name__)
app.debug = True
app.config['SECRET_TYPE'] = os.environ.get('SECRET')
app.config['SQLALCHEMY_DATABASE_URI'] = "sqlite:///chat.db"
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
app.config['SESSION_TYPE'] = "sqlalchemy"

db = SQLAlchemy(app)
app.config['SESSION_SQLALCHEMY'] = db

Session(app)

socketio = SocketIO(app, manage_session=False)

@app.route("/", methods=['GET', 'POST'])
def hello_world():
    return render_template('index.html')

@app.route("/chat", methods=['GET', 'POST'])
def chat():
    if request.method == 'POST':
        username = request.form['username']
        room = request.form['room']
        session['username'] = username
        session['room'] = room
        return render_template('chat.html', session = session)
    else:
        if(session.get('username') is not None):
            return render_template('chat.html', session = session)
        else:
            return redirect(url_for('index'))
        
@socketio.on('join', namespace='/chat')
def join(message):
    room = session.get('room')
    join_room(room)
    emit('status', {'msg': session.get('username') + 'has enterted the room.'}, room = room)

@socketio.on('text', namespace='/chat')
def text(message):
    room = session.get('room')
    emit('message', {'msg': session.get('username') + ' : ' + message['msg']}, room=room)

@socketio.on('left', namespace='/chat')
def left(message):
    room = session.get('room')
    username = session.get('username')
    leave_room(room)
    session.clear()
    emit('status', {'msg': username + ' has left the room.'}, room=room)

if __name__ == '__main__':
    app.run()`

chat.html:
`<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="Content-Security-Policy" content="upgrade-insecure-requests"> 
    <title>Chat</title>
    <link rel="stylesheet" href="../static/style.css">
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">

    <script src="https://code.jquery.com/jquery-3.6.0.min.js"
        integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous">
        </script>
    <script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/4.4.0/socket.io.min.js"></script>
    <script type="text/javascript">
        var socket;
        $(document).ready(function () {
            socket = io.connect('http://' + document.domain + ':' + location.port + '/chat');
            socket.on('connect', function () {
                socket.emit('join', {});
            });
            socket.on('status', function (data) {
                $('#chat').val($('#chat').val() + '<' + data.msg + '>\n');
                $('#chat').scrollTop($('#chat')[0].scrollHeight);
            });
            socket.on('message', function (data) {
                $('#chat').val($('#chat').val() + data.msg + '\n');
                $('#chat').scrollTop($('#chat')[0].scrollHeight);
            });
            $('#send').click(function (e) {
                text = $('#text').val();
                $('#text').val('');
                socket.emit('text', { msg: text });
            });
        });
        function leave_room() {
            socket.emit('left', {}, function () {
                socket.disconnect();
                // go back to the login page
                window.location.href = "{{ url_for('hello_world') }}";
            });
        }
    </script>

</head>

<body class="text-center">
    <div class="chatwindow">
        <h2>CET ChatBox</h2>
        <h2>Room: {{session['room']}}</h2><br>
        <textarea id="chat" cols="70" rows="10" placeholder="No messages yet."></textarea><br>
        <input type="text" id="text" size="60" placeholder="Enter your message here">
        <button type="button" id="send" class="btn btn-success">SEND</button><br><br>
        <center><button type="button" class="btn btn-danger" onclick=leave_room()>Leave</button></center>
    </div>
</body>

</html>`
Any help will be great as to how to solve this issue
Requirements.txt:
bidict==0.21.4
cachelib==0.4.1
certifi==2021.10.8
cffi==1.15.0
charset-normalizer==2.0.8
click==8.0.3
colorama==0.4.4
dnspython==2.1.0
eventlet==0.33.0
Flask==2.0.2
Flask-Session==0.4.0
Flask-SocketIO==5.1.1
Flask-SQLAlchemy==2.5.1
gevent==21.8.0
gevent-websocket==0.10.1
greenlet==1.1.2
gunicorn==20.1.0
idna==3.3
itsdangerous==2.0.1
Jinja2==3.0.3
MarkupSafe==2.0.1
pycparser==2.21
python-engineio==4.3.0
python-socketio==5.5.0
requests==2.26.0
six==1.16.0
SQLAlchemy==1.4.27
urllib3==1.26.7
Werkzeug==2.0.2
zope.event==4.5.0
zope.interface==5.4.0

**Comment by darrachequesne** [comment]
Hi! Did you check the scalability notes here: https://python-socketio.readthedocs.io/en/latest/server.html#scalability-notes?
